### PR TITLE
Lazify more work in builtins targets

### PR DESCRIPTION
**Pull request by Urgau**
This PR make some fields in `Target` and `TargetOptions` lazy.

This is done in order to reduce the impact of check-cfg which needs to load all the built-ins targets but doesn't need all the fields.

In particular this PR introduce a `MaybeLazy<T>` struct to support a borrowed, owned and lazy state.

~~The fields that are changed are:~~
 - ~~`LinkArgs` fields (access to env + allocate): 54 023 236 ins[^1] -> 53 478 072 ins~~
 - ~~`CrtObjects` fields (allocate): 53 478 072 ins -> 53 127 832 ins~~
 - ~~`llvm_name` field (access to env + allocate): 53 127 832 ins -> 53 057 110 ins~~

~~This brings around -1 000 000 ins*tructions* and -0.2/0.3ms of less wall-time (with some measurement even having the same minimum wall-time with or without check-cfg) :tada:.~~

See the latest perf run for the actual improvements, https://github.com/rust-lang/rust/pull/122703#issuecomment-2097100875.

*This PR is also a step further to completely `static`-fying all the build-in targets, if we one day we decide to do it, but that's a separate conversion.*

[^1]: This is the total number of instructions as measured with `cachegrind` for the entire `rustc` invocation on a cargo --lib hello world, variance was really low (<15 000 ins).
In the scale of the check-cfg feature, last time I measured `CheckCfg::fill_well_known` was around ~2.8 millions ins.

r? @petrochenkov

### CODE REVIEW

**Comment by Nilstrieb** [comment]
0.2ms is not a lot, even if you amortize it across hundreds of rustc invocations, and this does make the code more complicated. Is this really worth it?

**Comment by Urgau** [comment]
Actually, now that I remeasured it, the difference is big than I previously though. With stage1 the minimum wall-time measured is now the same, while before with nightly (so way more aggressive optimization) the minimum was never below 0.4ms; therefore this PR has the potential to *nearly eliminate* the cost of loading all the well known cfgs.

So yes, I definitively think it's worth it.

<details>

```text
$ hyperfine --warmup 50 "rustc +stage1 --crate-type=lib src/lib.rs" "rustc +stage1 --crate-type=lib --check-cfg='cfg()' -Zunstable-options src/lib.rs"
Benchmark 1: rustc +stage1 --crate-type=lib src/lib.rs
  Time (mean ± σ):      20.3 ms ±   1.5 ms    [User: 11.3 ms, System: 8.0 ms]
  Range (min … max):    17.9 ms …  24.9 ms    132 runs

Benchmark 2: rustc +stage1 --crate-type=lib --check-cfg='cfg()' -Zunstable-options src/lib.rs
  Time (mean ± σ):      20.7 ms ±   1.7 ms    [User: 12.7 ms, System: 6.9 ms]
  Range (min … max):    17.9 ms …  25.5 ms    133 runs

Summary
  rustc +stage1 --crate-type=lib src/lib.rs ran
    1.02 ± 0.11 times faster than rustc +stage1 --crate-type=lib --check-cfg='cfg()' -Zunstable-options src/lib.rs
```
```text
$ hyperfine --warmup 50 "rustc +nightly --crate-type=lib src/lib.rs" "rustc +nightly --crate-type=lib --check-cfg='cfg()' -Zunstable-options src/lib.rs"
Benchmark 1: rustc +nightly --crate-type=lib src/lib.rs
  Time (mean ± σ):      17.8 ms ±   1.3 ms    [User: 8.2 ms, System: 8.6 ms]
  Range (min … max):    15.8 ms …  26.5 ms    159 runs

Benchmark 2: rustc +nightly --crate-type=lib --check-cfg='cfg()' -Zunstable-options src/lib.rs
  Time (mean ± σ):      18.3 ms ±   1.3 ms    [User: 9.1 ms, System: 8.0 ms]
  Range (min … max):    16.3 ms …  25.5 ms    148 runs

Summary
  rustc +nightly --crate-type=lib src/lib.rs ran
    1.03 ± 0.10 times faster than rustc +nightly --crate-type=lib --check-cfg='cfg()' -Zunstable-options src/lib.rs
```

</details>

**Comment by petrochenkov** [comment]
I dislike this for the same reason as https://github.com/rust-lang/rust/pull/122207, but at much larger scale.
I'd also prefer to not do this, unless cfg checking is enabled by default, blocking on https://github.com/rust-lang/cargo/pull/13571.
Maybe be we'll *have to* do this, but it won't be a good day for target specs.
@rustbot blocked

**Comment by Urgau** [comment]
> I dislike this for the same reason as #122207, but at much larger scale.

I understand why you don't like https://github.com/rust-lang/rust/pull/122207, I also don't like it much, I mainly proposed that one for discussion and perf test.

However I think this PR is quite different, it goes into the direction of static target definitions, which IMO is sufficient in it self, even if it doesn't bring any perf improvements. @petrochenkov  Could you elaborate on why you don't like this PR?

**Comment by Urgau** [comment]
https://github.com/rust-lang/cargo/pull/13571 has been merged, the PR has been rebased and the [latest perf run](https://github.com/rust-lang/rust/pull/122703#issuecomment-2097100875) shows some clear improvements across multiple benchmarks :tada:.

@rustbot labels -S-blocked +S-waiting-on-review

**Comment by petrochenkov** [comment]
@Urgau 
To clarify the status, the default cfg checking is now in both stable and bootstrap rustc and cargo, and is not going to be reverted due to issues like https://github.com/rust-lang/rust/issues/124800 and similar, right?

File: compiler/rustc_target/src/spec/maybe_lazy.rs [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+    ///
+    /// This function pointer cannot *ever* take a closure. User can potentially
+    /// workaround that by using closure-to-fnptr or `const` items.
+    #[inline]
+    pub const fn lazy(a: fn() -> <T as ToOwned>::Owned) -> Self {
+        Self { inner: MaybeLazyInner::Lazy(LazyLock::new(a)) }
+    }
+}
+
+impl<T: 'static + ?Sized + ToOwned> Clone for MaybeLazy<T>
```
Comment (petrochenkov) [inline compiler/rustc_target/src/spec/maybe_lazy.rs:51]:
```suggestion
impl<T: 'static + ?Sized + ToOwned<Owned: Clone>> Clone for MaybeLazy<T>
```
In general, all `where` clauses below look convertible to simple bounds on parameters.

File: compiler/rustc_target/src/spec/maybe_lazy.rs [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+        Self {
+            inner: MaybeLazyInner::Cow(match &self.inner {
+                MaybeLazyInner::Lazy(f) => Cow::Owned((*f).to_owned()),
+                MaybeLazyInner::Cow(c) => c.clone(),
+            }),
+        }
+    }
+}
+
+impl<T: 'static + ?Sized + ToOwned> Deref for MaybeLazy<T>
```
Comment (petrochenkov) [inline compiler/rustc_target/src/spec/maybe_lazy.rs:66]: Are all other methods below implemented through this `Deref`?
Worth leaving a comment about that.

File: compiler/rustc_target/src/spec/maybe_lazy.rs [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+
+impl<T: 'static + ?Sized + ToOwned + Display> Display for MaybeLazy<T>
+where
+    <T as ToOwned>::Owned: Display,
+{
+    #[inline]
+    fn fmt(&self, fmt: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match &self.inner {
+            MaybeLazyInner::Lazy(f) => Display::fmt(&**f, fmt),
+            MaybeLazyInner::Cow(c) => Display::fmt(c, fmt),
```
Comment (petrochenkov) [inline compiler/rustc_target/src/spec/maybe_lazy.rs:102]: Why aren't `Debug` and `Display` implemented in terms of `Deref`?

File: compiler/rustc_target/src/spec/maybe_lazy.rs [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+    // make this struct usable to statics and we might need to
+    // workaround const-eval limitation (particulary around drop).
+    inner: MaybeLazyInner<T>,
+}
+
+impl<T: 'static + ?Sized + ToOwned> MaybeLazy<T> {
+    /// Create a [`MaybeLazy`] from an borrowed `T`.
+    #[inline]
+    pub const fn borrowed(a: &'static T) -> Self {
+        Self { inner: MaybeLazyInner::Cow(Cow::Borrowed(a)) }
```
Comment (petrochenkov) [inline compiler/rustc_target/src/spec/maybe_lazy.rs:32]:
```suggestion
        MaybeLazy { inner: MaybeLazyInner::Cow(Cow::Borrowed(a)) }
```
Could you avoid using `Self` in non-generic code? It becomes hard to search code (e.g. where something is constructed).

File: compiler/rustc_target/src/spec/mod.rs [hunk 5]
```diff
@@ -1830,7 +1833,7 @@ impl TargetWarnings {
 #[derive(PartialEq, Clone, Debug)]
 pub struct Target {
     /// Target triple to pass to LLVM.
-    pub llvm_target: StaticCow<str>,
+    pub llvm_target: MaybeLazy<str>,
```
Comment (petrochenkov) [inline compiler/rustc_target/src/spec/mod.rs:1836]: What is the policy for converting `StaticCow`s to `MaybeLazy`s?
I see there are many `StaticCow`s remaining.

Only fields initialized to non-literals in built-in targets are converted?

File: compiler/rustc_target/src/spec/mod.rs [hunk 6]
```diff
@@ -2345,24 +2348,20 @@ impl TargetOptions {
         link_args
     }
 
-    fn add_pre_link_args(&mut self, flavor: LinkerFlavor, args: &[&'static str]) {
```
Comment (petrochenkov) [inline compiler/rustc_target/src/spec/mod.rs:2348]: Why was this removed?

File: compiler/rustc_target/src/spec/mod.rs [hunk 7]
```diff
@@ -1965,34 +1968,34 @@ pub struct TargetOptions {
     linker_is_gnu_json: bool,
 
     /// Objects to link before and after all other object code.
-    pub pre_link_objects: CrtObjects,
-    pub post_link_objects: CrtObjects,
+    pub pre_link_objects: MaybeLazy<CrtObjects>,
```
Comment (petrochenkov) [inline compiler/rustc_target/src/spec/mod.rs:1971]: `LinkArgs` and `CrtObjects` are always supposed to be lazy, so `MaybeLazy` can be moved into the type alias themselves.

File: compiler/rustc_target/src/spec/mod.rs [hunk 8]
```diff
@@ -2452,20 +2452,20 @@ impl Default for TargetOptions {
             static_position_independent_executables: false,
             plt_by_default: true,
             relro_level: RelroLevel::None,
-            pre_link_objects: Default::default(),
-            post_link_objects: Default::default(),
-            pre_link_objects_self_contained: Default::default(),
-            post_link_objects_self_contained: Default::default(),
+            pre_link_objects: MaybeLazy::lazy(Default::default),
```
Comment (petrochenkov) [inline compiler/rustc_target/src/spec/mod.rs:2455]: Maybe keep the `Default` impls?
I don't see why this needs to be changed now, if we ever get to constructing targets at compile time, const traits should also be ready at that point.

File: compiler/rustc_target/src/spec/base/aix.rs [hunk 9]
```diff
@@ -22,10 +23,12 @@ pub fn opts() -> TargetOptions {
         is_like_aix: true,
         default_dwarf_version: 3,
         function_sections: true,
-        pre_link_objects: crt_objects::new(&[
```
Comment (petrochenkov) [inline compiler/rustc_target/src/spec/base/aix.rs:25]: The lazy part can be moved to `crt_objects::new` if CRT objects are always lazy.

File: compiler/rustc_target/src/spec/base/avr_gnu.rs [hunk 10]
```diff
@@ -24,11 +23,12 @@ pub fn target(target_cpu: &'static str, mmcu: &'static str) -> Target {
 
             linker: Some("avr-gcc".into()),
             eh_frame_header: false,
-            pre_link_args: TargetOptions::link_args(LinkerFlavor::Gnu(Cc::Yes, Lld::No), &[mmcu]),
-            late_link_args: TargetOptions::link_args(
-                LinkerFlavor::Gnu(Cc::Yes, Lld::No),
-                &["-lgcc"],
-            ),
+            pre_link_args: MaybeLazy::lazy(|| {
+                TargetOptions::link_args(LinkerFlavor::Gnu(Cc::Yes, Lld::No), &["-mmcu=atmega328"])
```
Comment (petrochenkov) [inline compiler/rustc_target/src/spec/base/avr_gnu.rs:27]: The lazy part can be moved to `TargetOptions::link_args` (and similar functions if any) if link args are always lazy.

File: compiler/rustc_target/src/spec/base/apple/mod.rs [hunk 11]
```diff
@@ -151,7 +156,7 @@ pub fn opts(os: &'static str, arch: Arch, abi: TargetAbi) -> TargetOptions {
         // macOS has -dead_strip, which doesn't rely on function_sections
         function_sections: false,
         dynamic_linking: true,
-        pre_link_args: pre_link_args(os, arch, abi),
+        pre_link_args,
```
Comment (petrochenkov) [inline compiler/rustc_target/src/spec/base/apple/mod.rs:159]: Why was this moved outside of `opts`?
I think we can still create lazy args here (especially if `fn pre_link_args` already returns lazy in the first place).

File: compiler/rustc_target/src/spec/maybe_lazy.rs (hunk 2 above)
Comment (Urgau) [reply compiler/rustc_target/src/spec/maybe_lazy.rs]: Some of them were indeed implemented trough the `Deref`. I've added a comment about that.

File: compiler/rustc_target/src/spec/maybe_lazy.rs (hunk 3 above)
Comment (Urgau) [reply compiler/rustc_target/src/spec/maybe_lazy.rs]: Because I had some trouble with them at first, but it seems to be working fine now. Switch to using it.

File: compiler/rustc_target/src/spec/maybe_lazy.rs (hunk 4 above)
Comment (Urgau) [reply compiler/rustc_target/src/spec/maybe_lazy.rs]: Done. (I'm wondering if we should have a lint about that :thinking:)

File: compiler/rustc_target/src/spec/mod.rs (hunk 5 above)
Comment (Urgau) [reply compiler/rustc_target/src/spec/mod.rs:1880]: I have been fairly conservative about changing fields to use `MaybeLazy`, mainly because I didn't feel changing everything without first discussing about it but also because I don't think it's necessary.

The policy I followed was to change fields may that allocate memory or change their value based on some environment variable, like `llvm_name` (env access with apple targets) or the link args (`BTreeMap` allocation by many targets), ...

But many fields are static on all targets and don't change, so I don't see the benefits of changing them.

**Comment by Urgau** [comment]
Regarding https://github.com/rust-lang/rust/pull/122703#discussion_r1617710772, https://github.com/rust-lang/rust/pull/122703#discussion_r1617723945, https://github.com/rust-lang/rust/pull/122703#discussion_r1617726118 and https://github.com/rust-lang/rust/pull/122703#discussion_r1617744482.

[`LazyLock`](https://doc.rust-lang.org/nightly/std/sync/struct.LazyLock.html) is defined as like this:
```rust
pub struct LazyLock<T, F = fn() -> T> { /* private fields */ }
```
the `F` generic parameter is the lazy **function pointer** that is called the first time to initialize the lazy lock.

A closure can only be converted into a function pointer when it doesn't capture anything from the environment (no args, no variables).

That means that a hypothetical `TargetOptions::link_args` function defined like this:
```rust
impl TargetOptions {
    pub fn link_args(flavor: LinkerFlavor, args: &[&str]) -> MaybeLazy<LinkArgs> {
        MaybeLazy::lazy(|| {
            let mut link_args = LinkArgs::new();
            add_link_args(&mut link_args, flavor, args);
            //                            ^^^^^^  ^^^^
            //                            captured from the environement (ie outside the closure)
            //                            preventing the closure from being converted into a
            //                            function pointer
            link_args
        })
    }
}
```
cannot work as the closure cannot be implicitly converted into function pointer.

The solution to this problem is to move out the `MaybeLazy::lazy(|| TargetOptions::link_args(...))` so that the arguments are always inside closure. That is what I have painfully done for all the link args, crt objects and some llvm_name fields.

One exception to the closure to fn ptr rule is const, including const generics but I excluded doing it as it creates some awkward functions and the `generic_const_exprs` feature is still an incomplete feature.

If you have a better way to deal with this, I'm all hear.

**Comment by Urgau** [comment]
> To clarify the status, the default cfg checking is now in both stable and bootstrap rustc and cargo, and is not going to be reverted due to issues like #124800 and similar, right?

As of 30-04-2024, the `rustc --check-cfg` flag is in beta and going to reach stable in 2 two weeks. Nothing is preventing it from reaching stable, the flag and the lint are working as expected.

Regarding Cargo, the always-on behavior is currently on nightly (since nightly-2024-05-05) and is going to reach stable by Rust 1.80 (in about 8 weeks from now). #124800 complain was about build-script requirement which was resolved in Cargo with the [`check-cfg` lint config](https://blog.rust-lang.org/2024/05/06/check-cfg.html#expecting-custom-cfgs). I see no reason to believe that the stabilization in Cargo is going to be reverted.

> Could you give a link to the PR with perf regressions that first enables cfg checking in a way observable to perf checking on merge?

Sure, https://github.com/rust-lang/rust/pull/124684#issuecomment-2094055103 is the Cargo subtree update where all the regressions can be seen. Unfortunately the perf result is full of outliers (ex. `syn` with it's 2000 warnings for example, which make the regression mostly about the lint machinery and not check-cfg it-self).

However it can still be determined that the regression on helloworld (probably the most realistic benchmark for check-cfg) is were up by ~6% and with this PR are down by ~2%.
This indicates that doing this optimization has a non-negliable benefit and that there are other areas of improvements (symbol locking, hashmap and vec allocations, ...).

@rustbot ready

File: compiler/rustc_target/src/spec/maybe_lazy.rs (hunk 4 above)
Comment (petrochenkov) [reply compiler/rustc_target/src/spec/maybe_lazy.rs]: This may be controversial, because a noticeable portion of people just use `Self` everywhere they can.
It may fly as a restriction lint in clippy.

File: compiler/rustc_target/src/spec/maybe_lazy.rs [hunk 12]
```diff
@@ -0,0 +1,125 @@
...
+}
+
+impl<T: 'static + ?Sized + ToOwned<Owned: Default>> Default for MaybeLazy<T> {
+    #[inline]
+    fn default() -> MaybeLazy<T> {
+        MaybeLazy::lazy(T::Owned::default)
+    }
+}
+
+// Debug and Display below are implemented in terms of this Deref
```
Comment (petrochenkov) [inline compiler/rustc_target/src/spec/maybe_lazy.rs:70]:
```suggestion
// Debug, Display and other traits below are implemented in terms of this Deref
```

**Comment by petrochenkov** [comment]
>If you have a better way to deal with this, I'm all hear.

I'll look tomorrow.

**Comment by petrochenkov** [comment]
@Urgau 
I think refactoring things into [this direction](https://github.com/petrochenkov/rust/commit/3ce60659ea34b6a78e93ca8dd649a8569973d154) as much as possible would be an improvement.
(Or switch `XxxData, Xxx` to `Xxx, LazyXxx` if the latter is used more often.)
@rustbot author

**Comment by Urgau** [comment]
> I think refactoring things into [this direction](https://github.com/petrochenkov/rust/commit/3ce60659ea34b6a78e93ca8dd649a8569973d154) as much as possible would be an improvement.

Good idea, I did that for almost all of the `CrtObjects` usage.

Regarding the link args, I couldn't do that this simply, since we can't captures arguments but I tried adding almost-identical `LazyLock` called `LazyLockState` that would allow `fn(S) -> T` with `S` being some state, which is useful because that would mean that `TargetOptions::link_args` could directly return a `MaybeLazy` and not have every targets import and use `MaybeLazy::lazy` to set a some link args.

Unfortunately I ran into some compiler errors regarding `DynSend` not being implemented for it and couldn't figure out how to solve that. You can see that attempt in [this branch](https://github.com/Urgau/rust/tree/lazy-targets-with-state).

@rustbot ready

**Comment by petrochenkov** [comment]
Standard `LazyLock` already supports stateful fn objects.
I was able to avoid all `MaybeLazy::lazy` calls for `CrtObjects` this way - https://github.com/rust-lang/rust/commit/2eefb88c5b90005d736bff91a4b35a96eeb4ebd4.

I suspect that most link args can be constructed this way too (from a static slice).
In other words, `TargetOptions::link_args` can work in target specs without modifications.

**Comment by petrochenkov** [comment]
Also, empty `BTreeMap`s are cheap to construct (no allocations), so they can be immediately created as owned, without laziness.
This applies to both crt objects and link args.

**Comment by Urgau** [comment]
for sanity check, let's verify that the latest changes didn't regress the perf improvements

@bors try @rust-timer queue

**Comment by Urgau** [comment]
hum, perf is neutral, that's unexpected, I'm still seeing reasonable instructions improvements locally; I see that I'm missing some `#[inline]` maybe that's hurting the perf?

@bors try @rust-timer queue

**Comment by Urgau** [comment]
I talked with `@Kobzol` and it seems like the regressions seen in `deep-vector` and `tt-muncher` are probably codegen-threshold reshuffle, making the regressions kind-of-FP.

Which leads to the perf improvements that are not as big as the first version. I didn't see anything obvious with `callgrind`; well except that making `Default` lazy seems to be more positive than I though (probably due to all the the drop). Let's verify this.

@bors try @rust-timer queue

File: compiler/rustc_target/src/spec/base/avr_gnu.rs (hunk 10 above)
Comment (Urgau) [reply compiler/rustc_target/src/spec/base/avr_gnu.rs]: `TargetOptions::link_args` has now returns a `MaybeLazy`.

File: compiler/rustc_target/src/spec/base/apple/mod.rs (hunk 11 above)
Comment (Urgau) [reply compiler/rustc_target/src/spec/base/apple/mod.rs]: `fn pre_link_args` now returns lazy, this is no longer the case.

**Comment by Urgau** [comment]
> Standard `LazyLock` already supports stateful fn objects. I was able to avoid all `MaybeLazy::lazy` calls for `CrtObjects` this way - [2eefb88](https://github.com/rust-lang/rust/commit/2eefb88c5b90005d736bff91a4b35a96eeb4ebd4).

Indeed. This has now been done for CRT objects and link args. *(pushed a separate commits, for easier review and maintenance for me)*

------

The latest (and most accurate) [perf run](https://github.com/rust-lang/rust/pull/122703#issuecomment-2173721529) doesn't show as much perf improvement as the [first one](https://github.com/rust-lang/rust/pull/122703#issuecomment-2097100875), I tried to improve more but it didn't lead anywhere.

(it's worth noting that they are not directly comparable given that they are more than a month apart - maybe worth redoing a perf run without the latest changes?)

@rustbot ready

File: compiler/rustc_target/src/spec/maybe_lazy.rs [hunk 13]
```diff
@@ -0,0 +1,144 @@
...
+pub struct MaybeLazy<T: 'static + ToOwned + ?Sized, F = fn() -> <T as ToOwned>::Owned> {
+    // Inner state.
+    //
+    // Not to be inlined since we may want in the future to
+    // make this struct usable to statics and we might need to
+    // workaround const-eval limitation (particulary around drop).
+    inner: MaybeLazyInner<T, F>,
+}
+
+impl<T: 'static + ?Sized + ToOwned, F: FnOnce() -> T::Owned> MaybeLazy<T, F> {
```
Comment (petrochenkov) [inline compiler/rustc_target/src/spec/maybe_lazy.rs:28]: Technicall the `FnOnce` bound is unnecessary for the `borrowed` and `owned` methods.

Its removal should make some other bounds below, e.g. in `From<&'static str>`, unnecessary.

File: compiler/rustc_target/src/spec/maybe_lazy.rs [hunk 14]
```diff
@@ -0,0 +1,144 @@
...
+    pub const fn lazied(f: F) -> Self {
+        MaybeLazy { inner: MaybeLazyInner::Lazy(LazyLock::new(f)) }
+    }
+}
+
+impl<T: 'static + ?Sized + ToOwned> MaybeLazy<T> {
+    /// Create a [`MaybeLazy`] from a function pointer.
+    #[inline]
+    pub const fn lazy(a: fn() -> T::Owned) -> Self {
+        MaybeLazy { inner: MaybeLazyInner::Lazy(LazyLock::new(a)) }
```
Comment (petrochenkov) [inline compiler/rustc_target/src/spec/maybe_lazy.rs:52]:
```suggestion
        Self::lazied(f)
```

File: compiler/rustc_target/src/spec/maybe_lazy.rs [hunk 15]
```diff
@@ -0,0 +1,144 @@
...
+            }),
+        }
+    }
+}
+
+impl<T: 'static + ?Sized + ToOwned<Owned: Default>, F: FnOnce() -> T::Owned> Default
+    for MaybeLazy<T, F>
+{
+    #[inline]
+    fn default() -> MaybeLazy<T, F> {
```
Comment (petrochenkov) [inline compiler/rustc_target/src/spec/maybe_lazy.rs:74]: Is this still necessary after optimizing the defaults to `Owned`?

File: compiler/rustc_target/src/spec/maybe_lazy.rs [hunk 16]
```diff
@@ -0,0 +1,144 @@
...
+impl<T: 'static + ?Sized + ToOwned, F: FnOnce() -> T::Owned> AsRef<T> for MaybeLazy<T, F> {
+    #[inline]
+    fn as_ref(&self) -> &T {
+        &**self
+    }
+}
+
+impl<
+    B: ?Sized + PartialEq<C> + ToOwned,
+    C: ?Sized + ToOwned,
```
Comment (petrochenkov) [inline compiler/rustc_target/src/spec/maybe_lazy.rs:121]: Maybe `T1`/`T2` instead of `B`/`C`?
Otherwise it's harder to see what is what.

File: compiler/rustc_target/src/spec/mod.rs [hunk 17]
```diff
@@ -2506,15 +2509,15 @@ impl Default for TargetOptions {
             pre_link_objects_self_contained: Default::default(),
             post_link_objects_self_contained: Default::default(),
             link_self_contained: LinkSelfContainedDefault::False,
-            pre_link_args: LinkArgs::new(),
+            pre_link_args: Default::default(),
```
Comment (petrochenkov) [inline compiler/rustc_target/src/spec/mod.rs:2512]: The default here still results in `Lazy`, rather than in empty `Owned`, from what I see.

File: compiler/rustc_target/src/spec/base/aix.rs [hunk 18]
```diff
@@ -1,5 +1,6 @@
 use crate::abi::Endian;
-use crate::spec::{crt_objects, cvs, Cc, CodeModel, LinkOutputKind, LinkerFlavor, TargetOptions};
+use crate::spec::TargetOptions;
+use crate::spec::{crt_objects, cvs, Cc, CodeModel, LinkOutputKind, LinkerFlavor};
```
Comment (petrochenkov) [inline compiler/rustc_target/src/spec/base/aix.rs:3]: Nit: unnecessary change.

**Comment by petrochenkov** [comment]
Let's maybe split these changes and try to benchmark and land them separately.
`MaybeLazy<str>` first, then CRT objects, then link args.
@rustbot author

File: compiler/rustc_target/src/spec/maybe_lazy.rs (hunk 13 above)
Comment (Urgau) [reply compiler/rustc_target/src/spec/maybe_lazy.rs:28]: I think it is, we are using `MaybeLazy::owned` with a `FnOnce`:
```
error[E0308]: mismatched types
    --> compiler/rustc_target/src/spec/mod.rs:3101:38
     |
2768 | /         macro_rules! key {
2769 | |             ($key_name:ident) => ( {
2770 | |                 let name = (stringify!($key_name)).replace("_", "-");
2771 | |                 if let Some(s) = obj.remove(&name).and_then(|s| s.as_str().map(str::to_strin...
...    |
3101 | |                     base.$key_name = MaybeLazy::owned(args);
     | |                     --------------   ^^^^^^^^^^^^^^^^^^^^^^ expected `MaybeLazy<BTreeMap<..., ...>, ...>`, found `MaybeLazy<_, fn() -> _>`
     | |                     |
     | |                     expected due to the type of this binding
...    |
3172 | |             } );
3173 | |         }
     | |_________- in this expansion of `key!`
...
3205 |           key!(post_link_objects = "post-link-objects", link_objects);
     |           ----------------------------------------------------------- in this macro invocation
     |
     = note: expected struct `MaybeLazy<BTreeMap<LinkOutputKind, Vec<Cow<'static, str>>>, LazyCrtObjectsState>`
                found struct `MaybeLazy<_, fn() -> _>`
```

File: compiler/rustc_target/src/spec/mod.rs (hunk 17 above)
Comment (petrochenkov) [reply compiler/rustc_target/src/spec/mod.rs:2512]: Ah, right, didn't notice.

File: compiler/rustc_target/src/spec/maybe_lazy.rs (hunk 15 above)
Comment (Urgau) [reply compiler/rustc_target/src/spec/maybe_lazy.rs:74]: I don't understand the question. What is "this" here?

This is the `Default` impl for `MaybeLazy`, it's the one called in `TargetOptions::default` with `Default::default`. It creates a `Owned` variant.

**Comment by petrochenkov** [comment]
https://github.com/rust-lang/rust/pull/127992 is closed so closing this as well.
